### ISSUE TITLE
ASP.NET Core "UseCorrectCatchAllBehavior" - impacts AutoRouteTransformer

### ISSUE THREAD

**Issue by rjpowers10** [OrchardCMS/OrchardCore#6926]
I've been struggling with an issue for a few days now and I figured I'd share what I found in case others are also struggling with it. There is a known bug in ASP.NET Core that can affect the AutoRoute module. Basically, depending on what routes your application has it's possible that the AutoRouteTransformer won't trigger. I was very confused why my AutoRoutes stopped working one day. To fix the issue you must opt in to the fix. It might be worth making a note of this in the AutoRoute doco.

[https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-3.1#route-template-reference](url)
![image](https://user-images.githubusercontent.com/68876423/90399348-d13edd00-e068-11ea-8cdb-546b1b5c6309.png)

For reference, this is how the AutoRouteTransformer is wired up
![image](https://user-images.githubusercontent.com/68876423/90399438-ee73ab80-e068-11ea-95a5-18a4c85d972a.png)

**Comment by sebastienros** [comment]
Can you provide repro steps so we can see if enabling the switch won't break anything. Which is why they made it a switch I assume.

**Comment by rjpowers10** [comment]
@sebastienros 

- Create a new site using the blog recipe
- Create new page using the route AutorouteDemo/MyPage
![image](https://user-images.githubusercontent.com/68876423/91479612-b42bba80-e86f-11ea-80ab-bad6f72d7cdb.png)
- Add a controller to the blog theme. I used
```
using Microsoft.AspNetCore.Mvc;

namespace TheBlogTheme.Controllers
{
    public class AutorouteDemoController : Controller
    {
        public string Index()
        {
            return "Hello world";
        }
    }
}
```
- Map the route in startup
```
using System;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Routing;
using Microsoft.Extensions.DependencyInjection;
using OrchardCore.Modules;
using OrchardCore.ResourceManagement;

namespace OrchardCore.Themes.TheBlogTheme
{
    public class Startup : StartupBase
    {
        public override void ConfigureServices(IServiceCollection serviceCollection)
        {
            serviceCollection.AddScoped<IResourceManifestProvider, ResourceManifest>();
        }

        public override void Configure(IApplicationBuilder app, IEndpointRouteBuilder routes, IServiceProvider serviceProvider)
        {
            routes.MapAreaControllerRoute(
                "AutoRouteDemo",
                "TheBlogTheme",
                "{controller}/{action}/{id?}");
        }
    }
}
```

- With the switch not set, I can invoke my controller method but not my page:
![image](https://user-images.githubusercontent.com/68876423/91479818-0a006280-e870-11ea-8f9b-e91aa8500b8f.png)

![image](https://user-images.githubusercontent.com/68876423/91479838-0ff64380-e870-11ea-954b-435c302d1352.png)

- By turning the switch on, both routes work
```
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using OrchardCore.Logging;

namespace OrchardCore.Cms.Web
{
    public class Program
    {
        public static Task Main(string[] args)
        {
            AppContext.SetSwitch("Microsoft.AspNetCore.Routing.UseCorrectCatchAllBehavior", true);
            return BuildHost(args).RunAsync();
        }

        public static IHost BuildHost(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureLogging(logging => logging.ClearProviders())
                .ConfigureWebHostDefaults(webBuilder => webBuilder
                    .UseStartup<Startup>()
                    .UseNLogWeb())
                .Build()
            ;
    }
}
```
![image](https://user-images.githubusercontent.com/68876423/91479892-269c9a80-e870-11ea-98b7-811df5a3586f.png)

**Comment by jtkech** [comment]
Yes, it has been fixed [through this PR](https://github.com/dotnet/aspnetcore/pull/20801), but only effective through an opt-in switch until 5.0, [see this PR](https://github.com/dotnet/aspnetcore/pull/21200) for more details.

**Comment by agriffard** [comment]
@jtkech Does this mean it will only be fixed when we switch to 5.0?

**Comment by jtkech** [comment]
Yes, or before but by using an opt-in switch
